File: frontend-staff/frontend_staff/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final int stock;
  final double price;
  final String category;
  final String? description;
  final String? sku;
  final bool isActive;
  final int lowStockThreshold;
  final DateTime createdAt;
  final DateTime updatedAt;

  Product({
    required this.id,
    required this.name,
    required this.stock,
    required this.price,
    required this.category,
    this.description,
    this.sku,
    required this.isActive,
    required this.lowStockThreshold,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'] as String,
      name: json['name'] as String,
      stock: json['stock'] as int,
      price: (json['price'] as num).toDouble(),
      category: json['category'] as String? ?? 'General',
      description: json['description'] as String?,
      sku: json['sku'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      lowStockThreshold: json['lowStockThreshold'] as int? ?? 5,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  bool get isOutOfStock => stock == 0;
  bool get isLowStock => stock > 0 && stock <= lowStockThreshold;

  String get stockStatus {
    if (isOutOfStock) return 'Out of Stock';
    if (isLowStock) return 'Low Stock';
    return 'In Stock';
  }

  String get formattedPrice => '₹${price.toStringAsFixed(0)}';
}

File: backend/models/Product.js
const mongoose = require('mongoose');

const productSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true,
        trim: true,
        unique: true
    },
    description: {
        type: String,
        trim: true,
        default: ''
    },
    price: {
        type: Number,
        required: true,
        min: 0
    },
    stock: {
        type: Number,
        required: true,
        min: 0,
        default: 0
    },
    category: {
        type: String,
        trim: true,
        default: 'General'
    },
    sku: {
        type: String,
        trim: true,
        unique: true,
        sparse: true
    },
    isActive: {
        type: Boolean,
        default: true
    },
    lowStockThreshold: {
        type: Number,
        default: 5,
        min: 0
    },
    createdAt: {
        type: Date,
        default: Date.now
    },
    updatedAt: {
        type: Date,
        default: Date.now
    }
});

// Update the updatedAt field before saving
productSchema.pre('save', function() {
    this.updatedAt = Date.now();
});

// Virtual for stock status
productSchema.virtual('stockStatus').get(function() {
    if (this.stock === 0) return 'out_of_stock';
    if (this.stock <= this.lowStockThreshold) return 'low_stock';
    return 'in_stock';
});

// Virtual for formatted price
productSchema.virtual('formattedPrice').get(function() {
    return `₹${this.price}`;
});

// Ensure virtual fields are serialized
productSchema.set('toJSON', { virtuals: true });

module.exports = mongoose.model('Product', productSchema);